src/fam.rs

```diff
@@ -0,0 +1,696 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+use std::mem::{self, size_of};
+
+/// Errors associated with the [`FamStructWrapper`](struct.FamStructWrapper.html) struct.
+#[derive(Clone, Debug)]
+pub enum Error {
+    /// The max size has been exceeded
+    SizeLimitExceeded,
+}
+
+/// Trait for accessing some properties of C defined FAM structures.
+///
+/// In C 99 an array of unknown size may appear within a struct definition as the last member
+/// (as long as there is at least one other named member).
+/// This is known as a flexible array member (FAM).
+/// Pre C99, the same behavior could be achieved using zero length arrays.
+///
+/// Flexible Array Members are the go-to choice for working with large amounts of data
+/// prefixed by header values.
+///
+/// For example the kvm API has many structures that are very similar to the `MockFamStruct`
+/// structure defined in the example bellow.
+///
+/// This is unsafe due to the number of constraints that aren't checked:
+/// * the implementer should be a POD
+/// * the implementor should contain a flexible array member of elements of type `Entry`
+/// * `Entry` should be a POD
+///
+/// Violating these may cause problems.
+///
+/// # Example
+///
+/// ```
+/// use vmm_sys_util::fam::*;
+///
+/// #[repr(C)]
+/// #[derive(Default)]
+/// pub struct __IncompleteArrayField<T>(::std::marker::PhantomData<T>, [T; 0]);
+/// impl<T> __IncompleteArrayField<T> {
+///     #[inline]
+///     pub fn new() -> Self {
+///         __IncompleteArrayField(::std::marker::PhantomData, [])
+///     }
+///     #[inline]
+///     pub unsafe fn as_ptr(&self) -> *const T {
+///         ::std::mem::transmute(self)
+///     }
+///     #[inline]
+///     pub unsafe fn as_mut_ptr(&mut self) -> *mut T {
+///         ::std::mem::transmute(self)
+///     }
+///     #[inline]
+///     pub unsafe fn as_slice(&self, len: usize) -> &[T] {
+///         ::std::slice::from_raw_parts(self.as_ptr(), len)
+///     }
+///     #[inline]
+///     pub unsafe fn as_mut_slice(&mut self, len: usize) -> &mut [T] {
+///         ::std::slice::from_raw_parts_mut(self.as_mut_ptr(), len)
+///     }
+/// }
```

Comment:
Could the __IncompleteArrayField struct be exported as a pub generic type, seems it only contains generic code.

Reply:
This struct is generated automatically by bindgen. You can find it for example in the kvm-bindings project. I needed to copy it just for the unit tests and the example. I don't know if it adds any value exporting it.